File: bindings/native/node/vibe.mjs
// vibe.mjs — VIBE as native JavaScript syntax.
//
// The `vibe` tagged template literal makes a VIBE document a first-class JS
// expression. You write the document inline — real template-literal syntax,
// with `${}` interpolation — and get back a plain, frozen JS object:
//
//     import { vibe } from "./vibe.mjs";
//
//     const port = 8080;
//     const cfg = vibe`
//         name    my-service
//         port    ${port}
//         tls     true
//         origins [ https://a.example  https://b.example ]
//         db {
//             host  localhost
//             port  5432
//         }
//     `;
//
//     cfg.name          // "my-service"   — native property access
//     cfg.port          // 8080           — a real JS number
//     cfg.db.host       // "localhost"    — nested, no getters
//     cfg.origins[0]    // "https://a.example"
//     for (const o of cfg.origins) ...   // it's a real Array
//
// Interpolated values are serialized to VIBE before splicing, so `${port}`,
// `${true}`, `${"a string"}`, and `${["x","y"]}` all Just Work and are safely
// quoted/escaped — no injection. A malformed document throws at evaluation.
//
// This is a TRUE native addon underneath: it #includes vibe.h and links
// libvibe.a (see vibe_addon.c). No FFI marshalling.
//
// SPDX-License-Identifier: MIT
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
const here = dirname(fileURLToPath(import.meta.url));
const native = require(join(here, "build", "Release", "vibe.node"));

export const version = () => native.version();

// Serialize an interpolated JS value into a VIBE fragment that lexes as one
// value. Strings are always quoted+escaped so nothing a user interpolates can
// break out into structure (no injection). Arrays become `[ a b c ]`.
function toVibe(x) {
  if (x === null || x === undefined) return '""';
  const t = typeof x;
  if (t === "number") return Number.isFinite(x) ? String(x) : '""';
  if (t === "boolean") return x ? "true" : "false";
  if (t === "bigint") return x.toString();
  if (Array.isArray(x)) return "[ " + x.map(toVibe).join(" ") + " ]";
  // Everything else: emit as a quoted, escaped VIBE string.
  const s = String(x)
    .replace(/\\/g, "\\\\")
    .replace(/"/g, '\\"')
    .replace(/\n/g, "\\n")
    .replace(/\t/g, "\\t")
    .replace(/\r/g, "\\r");
  return `"${s}"`;
}

// Parse VIBE source into a frozen native JS object.
export function parse(source) {
  const handle = native.parse(source); // throws on malformed input
  try {
    return deepFreeze(native.toObject(handle));
  } finally {
    native.free(handle);
  }
}

function deepFreeze(o) {
  if (o && typeof o === "object") {
    for (const k of Object.keys(o)) deepFreeze(o[k]);
    Object.freeze(o);
  }
  return o;
}

// The tagged template: `vibe`...`` is now valid, native VIBE-in-JS.
export function vibe(strings, ...values) {
  let src = strings[0];
  for (let i = 0; i < values.length; i++) {
    src += toVibe(values[i]) + strings[i + 1];
  }
  return parse(src);
}

// Also allow `vibe(stringVariable)` for documents built/loaded at runtime.
const tag = new Proxy(vibe, {
  apply(target, thisArg, args) {
    if (args.length === 1 && typeof args[0] === "string") return parse(args[0]);
    return Reflect.apply(target, thisArg, args);
  },
});

export default tag;

File: bindings/native/python/vibe.py
"""vibe — VIBE as native Python syntax.

Two ways to embed VIBE directly in Python, both backed by the native C
extension (``_vibe``, which #includes vibe.h and links libvibe.a):

1. The ``vibe`` callable — parse an inline document and get an object whose
   keys are attributes AND items::

       import vibe
       cfg = vibe('''
           name    my-service
           port    8080
           tls     true
           origins [ https://a.example  https://b.example ]
           db {
               host  localhost
               port  5432
           }
       ''')
       cfg.name          # "my-service"
       cfg.port          # 8080
       cfg.db.host       # "localhost"   (nested)
       cfg["db"]["port"] # 5432          (item access too)
       cfg.origins[0]    # "https://a.example"
       for o in cfg.origins: ...

2. The ``# coding: vibe`` source codec — makes ``vibe`` a real literal prefix
   so a whole module can carry VIBE blocks the interpreter decodes for you::

       # coding: vibe
       import vibe
       CONFIG = vibe'''
           name  my-service
           port  8080
       '''

   (register the codec once with ``vibe.install_codec()`` — see below.)

SPDX-License-Identifier: MIT
"""
from __future__ import annotations

import codecs
import encodings
import io
import re
import tokenize as _tok

import _vibe

VibeError = _vibe.VibeError

__all__ = ["vibe", "parse", "version", "VibeError", "View", "install_codec"]


def version() -> str:
    return _vibe.version()


class View:
    """Attribute- and item-accessible read-only view over a parsed VIBE value.

    Wraps the plain dict/list/scalars produced by the native extension so you
    can navigate with dots (``cfg.db.host``) or brackets (``cfg["db"]["host"]``)
    interchangeably. Missing keys raise the usual KeyError/AttributeError.
    """

    __slots__ = ("_d",)

    def __init__(self, data):
        object.__setattr__(self, "_d", data)

    @staticmethod
    def _wrap(v):
        if isinstance(v, dict):
            return View(v)
        if isinstance(v, list):
            return [View._wrap(x) for x in v]
        return v

    def __getattr__(self, name):
        d = object.__getattribute__(self, "_d")
        if name in d:
            return View._wrap(d[name])
        raise AttributeError(name)

    def __getitem__(self, key):
        return View._wrap(object.__getattribute__(self, "_d")[key])

    def __contains__(self, key):
        return key in object.__getattribute__(self, "_d")

    def __iter__(self):
        return iter(object.__getattribute__(self, "_d"))

    def keys(self):
        return object.__getattribute__(self, "_d").keys()

    def get(self, key, default=None):
        d = object.__getattribute__(self, "_d")
        return View._wrap(d[key]) if key in d else default

    def to_dict(self):
        return object.__getattribute__(self, "_d")

    def __len__(self):
        return len(object.__getattribute__(self, "_d"))

    def __repr__(self):
        return f"View({object.__getattribute__(self, '_d')!r})"

    def __eq__(self, other):
        if isinstance(other, View):
            other = other.to_dict()
        return object.__getattribute__(self, "_d") == other


def parse(source):
    """Parse VIBE text (str or bytes) into a navigable View.

    Raises vibe.VibeError on malformed input.
    """
    if isinstance(source, bytes):
        source = source.decode("utf-8")
    doc = _vibe.parse(source)
    return View._wrap(doc.to_dict())


# `vibe(...)` is the native-feeling entry point.
vibe = parse


# --------------------------------------------------------------------------
# Source codec: `# coding: vibe` turns `vibe'''...'''` / `vibe"""..."""` into
# `vibe(r'''...''')` so VIBE literals are a real part of the module's grammar.
# --------------------------------------------------------------------------
_utf8 = encodings.search_function("utf8")

# Match a `vibe` prefix immediately followed by a (possibly triple) quote.
_LITERAL = re.compile(
    r"""\bvibe(?P<q>'''|\"\"\"|'|")""",
)


def _rewrite(text: str) -> str:
    """Rewrite `vibe<quote>...<quote>` literals into `vibe(r<quote>...<quote>)`.

    Uses a small state machine so quotes inside the VIBE body don't confuse it:
    once we see an opening `vibe'''`, we scan to the matching closing quote.
    """
    out = []
    i = 0
    n = len(text)
    while i < n:
        m = _LITERAL.search(text, i)
        if not m:
            out.append(text[i:])
            break
        out.append(text[i:m.start()])
        q = m.group("q")
        body_start = m.end()
        end = text.find(q, body_start)
        if end == -1:
            # Unterminated — leave the rest untouched; Python will error clearly.
            out.append(text[m.start():])
            break
        body = text[body_start:end]
        # Wrap as a raw string call so backslashes/quotes pass through verbatim.
        # Guard against a raw-string ending in a backslash (illegal in Python).
        if body.endswith("\\"):
            body += " "
        out.append(f"vibe(r{q}{body}{q})")
        i = end + len(q)
    return "".join(out)


def _decode(input_bytes, errors="strict"):
    text, consumed = _utf8.decode(input_bytes, errors)
    return _rewrite(text), consumed


class _IncrementalDecoder(codecs.BufferedIncrementalDecoder):
    def _buffer_decode(self, input, errors, final):
        if final:
            return _decode(input, errors)
        return "", 0


def _search(name):
    if name.replace("-", "_") != "vibe":
        return None
    return codecs.CodecInfo(
        name="vibe",
        encode=_utf8.encode,
        decode=_decode,
        incrementaldecoder=_IncrementalDecoder,
        streamreader=_utf8.streamreader,
        streamwriter=_utf8.streamwriter,
    )


_installed = False


def install_codec():
    """Register the `# coding: vibe` source codec (idempotent)."""
    global _installed
    if not _installed:
        codecs.register(_search)
        _installed = True


# Auto-register on import so `# coding: vibe` works without a manual call.
install_codec()


# Make the *module itself* callable, so `import vibe; vibe("...")` works and the
# VIBE literal reads as native syntax (rather than requiring `vibe.vibe(...)`).
import sys as _sys


class _CallableModule(_sys.modules[__name__].__class__):
    def __call__(self, source):
        return parse(source)


_sys.modules[__name__].__class__ = _CallableModule
